### PR TITLE
[KYUUBI #5766] Default `spark.yarn.maxAppAttempts` to 1 for spark engine

### LINKED ISSUE #5766: [Bug][SPARK] kyuubi.engine.submit.time is wrong for ApplicationMaster retries
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the bug

```
ERROR ApplicationMaster: User class threw exception: org.apache.kyuubi.KyuubiException: The total engine initialization time (84513834 ms) exceeds kyuubi.session.engine.initialize.timeout (300000 ms), and submitted at 1700714879267.
org.apache.kyuubi.KyuubiException: The total engine initialization time (84513834 ms) exceeds kyuubi.session.engine.initialize.timeout (300000 ms), and submitted at 1700714879267.
	at org.apache.kyuubi.engine.spark.SparkSQLEngine$.main(SparkSQLEngine.scala:319)
	at org.apache.kyuubi.engine.spark.SparkSQLEngine.main(SparkSQLEngine.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
```

When Kyuubi spark engine starting failover, the kyuubi.engine.submit.time is wrong to use

### Affects Version(s)

1.7.0

### Kyuubi Server Log Output

_No response_

### Kyuubi Engine Log Output

_No response_

### Kyuubi Server Configurations

_No response_

### Kyuubi Engine Configurations

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to fix.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by wForget**
# :mag: Description
## Issue References 🔗

This pull request fixes #5766

## Describe Your Solution 🔧

As discussed in https://github.com/apache/kyuubi/issues/5766#issuecomment-1835285025, we should add `spark.yarn.maxAppAttempts=1` for spark engine when `spark.master` is `yarn`.

## Types of changes :bookmark:

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Related Unit Tests

---

# Checklists
## 📝 Author Self Checklist

- [x] My code follows the [style guidelines](https://kyuubi.readthedocs.io/en/master/contributing/code/style.html) of this project
- [x] I have performed a self-review
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

## 📝 Committer Pre-Merge Checklist

- [x] Pull request title is okay.
- [x] No license issues.
- [x] Milestone correctly set?
- [ ] Test coverage is ok
- [x] Assignees are selected.
- [x] Minimum number of approvals
- [x] No changes are requested

**Be nice. Be informative.**

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 1]
```diff
@@ -259,6 +260,15 @@ class SparkProcessBuilder(
     map.result().toMap
   }
 
+  def extraYarnConf: Map[String, String] = {
+    if (clusterManager().exists(_.toLowerCase(Locale.ROOT).startsWith("yarn"))) {
+      // set `spark.yarn.maxAppAttempts` to 1 to avoid invalid attempts.
+      Map(YARN_MAX_APP_ATTEMPTS_KEY -> "1")
```
Comment (cxzl25) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:266]: [YARN-5617](https://issues.apache.org/jira/browse/YARN-5617). AMs only intended to run one attempt can be run more than once


Fix Version/s: 2.9.0, 3.0.0-alpha4, 2.8.2

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:266]: maybe we should add such info to the comment

**Review by pan3793 — APPROVED** [review]

**Review by pan3793 — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala [hunk 2]
```diff
@@ -140,8 +140,9 @@ class SparkProcessBuilder(
       allConf = allConf ++ zkAuthKeytabFileConf(allConf)
     }
     // pass spark engine log path to spark conf
-    (allConf ++ engineLogPathConf ++ appendPodNameConf(allConf)).foreach { case (k, v) =>
-      buffer ++= confKeyValue(convertConfigKey(k), v)
+    (allConf ++ engineLogPathConf ++ extraYarnConf ++ appendPodNameConf(allConf)).foreach {
```
Comment (cxzl25) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala:143]: It looks like `extraYarnConf` will override the user's configuration.

Comment (wForget) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: I have moved extraYarnConf to the front

Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: Emm... if we really want to respect user's conf here, we need to do things like `appendPodNameConf` does

Comment (wForget) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: thanks, fixed

Comment (wForget) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/spark/SparkProcessBuilder.scala]: The construction of allConf is too complex, I prefer to put kvs into a mutable map, let's optimize it later.

**Review by pan3793 — APPROVED** [review]

**Review by cxzl25 — APPROVED** [review]

**Pull request closed without merging**